pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,581 @@
+	}
+
+	if kind != "" && kind != instancetype.SingularPreferenceResourceName && kind != instancetype.ClusterSingularPreferenceResourceName {
+		return flagErr(PreferenceFlag, "invalid preference kind: %s", kind)
+	}
+
+	// If kind is empty we rely on the vm-mutator to fill in the default value VirtualMachineClusterPreference
+	vm.Spec.Preference = &v1.PreferenceMatcher{
+		Name: name,
+		Kind: kind,
+	}
+
+	return nil
+}
+
+func withDataSourceVolume(c *createVM, vm *v1.VirtualMachine) error {
+	for _, dataSourceVol := range c.dataSourceVolumes {
+		vol := cloneVolume{}
+		err := mapParams(DataSourceVolumeFlag, dataSourceVol, &vol)
+		if err != nil {
+			return err
+		}
+
+		if vol.Source == "" {
+			return flagErr(DataSourceVolumeFlag, "src must be specified")
+		}
+
+		namespace, name, err := splitPrefixedName(vol.Source)
+		if err != nil {
+			return flagErr(DataSourceVolumeFlag, "src invalid: %w", err)
+		}
+
+		if vol.Name == "" {
+			vol.Name = fmt.Sprintf("%s-ds-%s", vm.Name, name)
+		}
+
+		dvt := v1.DataVolumeTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: vol.Name,
+			},
+			Spec: cdiv1.DataVolumeSpec{
+				Storage: &cdiv1.StorageSpec{},
+				SourceRef: &cdiv1.DataVolumeSourceRef{
+					Kind: "DataSource",
+					Name: name,
+				},
+			},
+		}
+		if namespace != "" {
+			dvt.Spec.SourceRef.Namespace = &namespace
+		}
+		if vol.Size != nil {
+			dvt.Spec.Storage.Resources.Requests = k8sv1.ResourceList{
+				k8sv1.ResourceStorage: *vol.Size,
+			}
+		}
+		vm.Spec.DataVolumeTemplates = append(vm.Spec.DataVolumeTemplates, dvt)
+
+		vm.Spec.Template.Spec.Volumes = append(vm.Spec.Template.Spec.Volumes, v1.Volume{
+			Name: vol.Name,
+			VolumeSource: v1.VolumeSource{
+				DataVolume: &v1.DataVolumeSource{
+					Name: vol.Name,
+				},
+			},
+		})
+	}
+
+	return nil
+}
+
+func withContainerdiskVolume(c *createVM, vm *v1.VirtualMachine) error {
+	for i, containerdiskVol := range c.containerdiskVolumes {
+		vol := cloneVolume{}
+		err := mapParams(ContainerdiskVolumeFlag, containerdiskVol, &vol)
+		if err != nil {
+			return err
+		}
+
+		if vol.Source == "" {
+			return flagErr(ContainerdiskVolumeFlag, "src must be specified")
+		}
+		if vol.Size == nil {
+			return flagErr(ContainerdiskVolumeFlag, "size must be specified")
+		}
+
+		if !strings.HasPrefix(vol.Source, "docker://") {
+			vol.Source = fmt.Sprintf("docker://%s", vol.Source)
+		}
+		if vol.Name == "" {
+			vol.Name = fmt.Sprintf("%s-containerdisk-%d", vm.Name, i)
+		}
+
+		vm.Spec.DataVolumeTemplates = append(vm.Spec.DataVolumeTemplates, v1.DataVolumeTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: vol.Name,
+			},
+			Spec: cdiv1.DataVolumeSpec{
+				Storage: &cdiv1.StorageSpec{
+					Resources: k8sv1.ResourceRequirements{
+						Requests: k8sv1.ResourceList{
+							k8sv1.ResourceStorage: *vol.Size,
+						},
+					},
+				},
+				Source: &cdiv1.DataVolumeSource{
+					Registry: &cdiv1.DataVolumeSourceRegistry{
+						URL: &vol.Source,
+					},
+				},
```

Comment:
for container disks, we have two use cases. People can use the containerDisk volumesource directly which results in an ephemeral volume that has writes that are erased after each VM restart. Then there's ability to import a containerdisk into a PVC in order to consume it as a persistent disk that survives restarts.

Maybe to avoid this confusion we could make a ```--volume-import``` that takes a src option that is either a container or an http endpoint. Then we construct the DVT to perform the import for us just like you have here.

Reply:
I agree but I would rather do this in a follow up PR. I changed `--volume-containerdisk` to use the ephemeral volumes for now .